tests/test_history.py

```diff
@@ -383,62 +458,61 @@ def hist_file():
     except FileNotFoundError:
         pass
 
-def test_existing_history_file(hist_file, capsys):
-    import atexit
-    import readline
-
-    # Create the history file before making cmd2 app
-    with open(hist_file, 'w'):
-        pass
-
-    # Create a new cmd2 app
-    cmd2.Cmd(persistent_history_file=hist_file)
-    _, err = capsys.readouterr()
-
-    # Make sure there were no errors
-    assert err == ''
-
-    # Unregister the call to write_history_file that cmd2 did
-    atexit.unregister(readline.write_history_file)
-
-    # Remove created history file
-    os.remove(hist_file)
-
-def test_new_history_file(hist_file, capsys):
-    import atexit
-    import readline
-
-    # Remove any existing history file
-    try:
-        os.remove(hist_file)
-    except OSError:
-        pass
-
-    # Create a new cmd2 app
-    cmd2.Cmd(persistent_history_file=hist_file)
-    _, err = capsys.readouterr()
-
-    # Make sure there were no errors
-    assert err == ''
-
-    # Unregister the call to write_history_file that cmd2 did
-    atexit.unregister(readline.write_history_file)
-
-    # Remove created history file
-    os.remove(hist_file)
-
 def test_bad_history_file_path(capsys, request):
-    # Use a directory path as the history file
-    test_dir = os.path.dirname(request.module.__file__)
+    with tempfile.TemporaryDirectory() as test_dir:
+        # Create a new cmd2 app
+        cmd2.Cmd(persistent_history_file=test_dir)
+        _, err = capsys.readouterr()
+        assert 'is a directory' in err
```

Comment:
This test is still failing on Windows/AppVeyor.  It looks like for that OS, it is not getting an `IsADirectoryError`, but an `OSError` instead.

One thing I've been thinking recently is that since we are now Python 3.4+, we should consider using [Pathlib](https://docs.python.org/3/library/pathlib.html) for paths and to do things like testing to make sure the path is a file.

Reply:
Yeah, I saw that it's still broken too. I'm already catching the `IsADirectoryError` exception, which is what shows the error message that appveyor is barfing on. This test was written to make sure that we generated an error if someone passed a directory as the persistent history file. For some reason, AppVeyor is throwing a permission error, not the `IsADirectoryError`. I'm working on it....

Reply:
OK, I'm stumped. `tempfile.TemporaryDirectory()` is supposed to create a directory that's readable and writable by the current user. It sure seems like it did, because I can create a new directory `somedir` inside of it. But when we go to `open(hist_file, 'rb')` either one of those directories, we get a permission error, not an `IsADirectoryError`.

Reply:
I'm spinning up my Windows VM to try to investigate.  This is weird.

Reply:
So on Windows when you try to open a directory in Python 3, you get an OSError with a message about permission denied.

Reply:
Got it. So instead of relying on the `IsADirectoryError` exception to tell me we are trying to open a directory, I need to manually check whether it's a directory using `os.is_dir()`. I'll try that.

Reply:
Yup you got it.  But `pathlib` has some advantages over `os`:  https://treyhunner.com/2018/12/why-you-should-be-using-pathlib/

We might want to experiment with using that because we can do things like:
```Python3
import pathlib
hist_file = pathlib.Path(hist_file).expanduser().resolve()
if hist_file.is_dir():
    # Handle case where the specified history file is actually a directory

if not hist_file.parent.is_dir()
    # Handle the problem that the directory the history file should be in isn't a directory
```

For Python 3.6 and newer you can use a `Path` directly in `open()`.  Before that you cannot, but you can cast to a `str` or use the `read_text` or `read_bytes` methods.

Reply:
It worked with the manual check if it's a directory, but I'm liking the `pathlib` approach, I didn't even know that thing existed. I'll switch it over from `os.path` to `path lib`.